modules/index/src/main/java/org/apache/ignite/internal/index/ChangeIndexStatusTaskScheduler.java

```diff
@@ -0,0 +1,214 @@
+import org.apache.ignite.internal.close.ManuallyCloseable;
+import org.apache.ignite.internal.cluster.management.topology.api.LogicalTopologyService;
+import org.apache.ignite.internal.hlc.HybridClock;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.network.ClusterService;
+import org.apache.ignite.internal.placementdriver.PlacementDriver;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+
+/** Component is responsible for starting and stopping {@link ChangeIndexStatusTask}. */
+class ChangeIndexStatusTaskScheduler implements ManuallyCloseable {
+    private static final IgniteLogger LOG = Loggers.forClass(ChangeIndexStatusTaskScheduler.class);
+
+    private final CatalogManager catalogManager;
+
+    private final ClusterService clusterService;
+
+    private final LogicalTopologyService logicalTopologyService;
+
+    private final HybridClock clock;
+
+    private final ClockWaiter clockWaiter;
+
+    private final PlacementDriver placementDriver;
+
+    private final Executor executor;
+
+    private final Map<ChangeIndexStatusTaskId, ChangeIndexStatusTask> taskById = new ConcurrentHashMap<>();
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean closeGuard = new AtomicBoolean();
+
+    ChangeIndexStatusTaskScheduler(
+            CatalogManager catalogManager,
+            ClusterService clusterService,
+            LogicalTopologyService logicalTopologyService,
+            HybridClock clock,
+            ClockWaiter clockWaiter,
+            PlacementDriver placementDriver,
+            Executor executor
+    ) {
+        this.catalogManager = catalogManager;
+        this.clusterService = clusterService;
+        this.logicalTopologyService = logicalTopologyService;
+        this.clock = clock;
+        this.clockWaiter = clockWaiter;
+        this.placementDriver = placementDriver;
+        this.executor = executor;
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        stopAllTasks();
+    }
+
+    /**
+     * Schedules a task for that will transfer the given index to the {@link CatalogIndexStatus#BUILDING} state.
+     */
+    void scheduleStartBuildingTask(CatalogIndexDescriptor indexDescriptor) {
+        assert indexDescriptor.status() == CatalogIndexStatus.REGISTERED;
+
+        LOG.info("Scheduling starting of index building. Index: {}", indexDescriptor);
+
+        inBusyLockSafe(busyLock, () -> {
+            var taskId = new ChangeIndexStatusTaskId(indexDescriptor);
+
+            scheduleTask(taskId, startBuildingIndexTask(indexDescriptor));
+        });
+    }
+
+    private ChangeIndexStatusTask startBuildingIndexTask(CatalogIndexDescriptor indexDescriptor) {
```

Comment:
I would add a postfix `Busy`.

Reply:
This is simply a shortcut for a constructor, I'm not sure it's worth mentioning the busy lock